lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,377 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package backend
+
+import (
+	"bytes"
+	"sync"
+
+	badger "github.com/dgraph-io/badger/v2"
+)
+
+const (
+	checkpointFlushMinSize = 1 << 20
+)
+
+func OpenBadger(path string) (Backend, error) {
+	opts := badger.DefaultOptions(path + ".badger") // temporary path safety
+	opts = opts.WithMaxCacheSize(64 << MiB).WithCompactL0OnClose(false)
+	opts.Logger = nil
+	bdb, err := badger.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+	return &badgerBackend{
+		bdb:    bdb,
+		closed: make(chan struct{}),
+	}, nil
+}
+
+func OpenBadgerMemory() Backend {
+	opts := badger.DefaultOptions("").WithInMemory(true)
+	opts.Logger = nil
+	bdb, err := badger.Open(opts)
+	if err != nil {
+		panic(err)
+	}
+	return &badgerBackend{
+		bdb:    bdb,
+		closed: make(chan struct{}),
+	}
+}
+
+// badgerBackend implements Backend on top of a badger
+type badgerBackend struct {
+	bdb     *badger.DB
+	closeWG sync.WaitGroup
+	closed  chan struct{}
+}
+
+func (b *badgerBackend) NewReadTransaction() (ReadTransaction, error) {
+	return badgerSnapshot{
+		txn: b.bdb.NewTransaction(false),
+		rel: newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *badgerBackend) NewWriteTransaction() (WriteTransaction, error) {
+	// We use two transactions here to preserve the property that our
+	// leveldb wrapper has, that writes in a transaction are completely
+	// invisible until it's committed, even inside that same transaction.
+	rtxn := b.bdb.NewTransaction(false)
+	wtxn := b.bdb.NewTransaction(true)
+	return &badgerTransaction{
+		badgerSnapshot: badgerSnapshot{
+			txn: rtxn,
+			rel: newReleaser(&b.closeWG),
+		},
+		txn: wtxn,
+		bdb: b.bdb,
+		rel: newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *badgerBackend) Close() error {
+	b.closeWG.Wait()
+	err := b.bdb.Close()
+	close(b.closed)
+	return wrapBadgerErr(err)
+}
+
+func (b *badgerBackend) Get(key []byte) ([]byte, error) {
+	txn := b.bdb.NewTransaction(false)
+	defer txn.Discard()
+	item, err := txn.Get(key)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	val, err := item.ValueCopy(nil)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	return val, nil
+}
+
+func (b *badgerBackend) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerPrefixIterator(txn, prefix)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) NewRangeIterator(first, last []byte) (Iterator, error) {
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerRangeIterator(txn, first, last)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) Put(key, val []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
```

Comment:
Are you not using `NewWriteTransaction` due to the additional read transaction created there? `Put` is probably mostly fast enough, but in principle you should also use the `closeWG` stuff here.

Reply:
Good call on the closeWG